testing/fil_token_integration/tests/frc46_multi_actor_tests.rs

```diff
@@ -0,0 +1,171 @@
+    let alice = tester.install_actor_stateless(TEST_ACTOR_WASM, 10010);
+    let bob = tester.install_actor_stateless(TEST_ACTOR_WASM, 10011);
+    let carol = tester.install_actor_stateless(TEST_ACTOR_WASM, 10012);
+    let dave = tester.install_actor_stateless(TEST_ACTOR_WASM, 10013);
+
+    // Instantiate machine
+    tester.instantiate_machine(DummyExterns).unwrap();
+
+    // construct actors
+    for actor in [token_actor, alice, bob, carol, dave] {
+        let ret_val = tester.call_method(operator[0].1, actor, method_hash!("Constructor"), None);
+        assert!(ret_val.msg_receipt.exit_code.is_success());
+    }
+
+    // TEST: alice sends bob a transfer of zero amount (rejecting first time and then accepting)
+    // first, tell bob to reject it
+    let params = action_params(token_actor, TestAction::Transfer(bob, action(TestAction::Reject)));
+    let ret_val = tester.call_method(operator[0].1, alice, method_hash!("Action"), Some(params));
+    // we told bob to reject, so the action call should return success but give us the error result as return data
+    assert!(ret_val.msg_receipt.exit_code.is_success());
+    // check the receipt we got in return data
+    let receipt = ret_val.msg_receipt.return_data.deserialize::<Receipt>().unwrap();
+    assert!(!receipt.exit_code.is_success());
+
+    // this time tell bob to accept it
+    let params = action_params(token_actor, TestAction::Transfer(bob, action(TestAction::Accept)));
+    let ret_val = tester.call_method(operator[0].1, alice, method_hash!("Action"), Some(params));
+    // we told bob to accept this time, so transfer should succeed
+    assert!(ret_val.msg_receipt.exit_code.is_success());
+
+    // balance should remain zero
+    let balance = tester.get_balance(operator[0].1, token_actor, alice);
+    assert_eq!(balance, TokenAmount::from_atto(0));
+    let balance = tester.get_balance(operator[0].1, token_actor, bob);
+    assert_eq!(balance, TokenAmount::from_atto(0));
+
+    // TEST: alice sends bob a transfer of a non-zero amounnt. As before, we'll reject it the first time then accept
+    // mint some tokens to alice first
+    let ret_val = tester.mint_tokens(
+        operator[0].1,
+        token_actor,
+        alice,
+        TokenAmount::from_atto(100),
+        action(TestAction::Accept),
+    );
+    assert!(ret_val.msg_receipt.exit_code.is_success());
+    let balance = tester.get_balance(operator[0].1, token_actor, alice);
+    assert_eq!(balance, TokenAmount::from_atto(100));
+    // now send to bob, who will reject them
+    let params = action_params(token_actor, TestAction::Transfer(bob, action(TestAction::Reject)));
+    let ret_val = tester.call_method(operator[0].1, alice, method_hash!("Action"), Some(params));
+    assert!(ret_val.msg_receipt.exit_code.is_success());
+    // check the receipt we got in return data
+    let receipt = ret_val.msg_receipt.return_data.deserialize::<Receipt>().unwrap();
+    assert!(!receipt.exit_code.is_success());
```

Comment:
Critical to assert all the balances here, and after every transfer.

Reply:
Think I've covered all these now.

Also added some extra checks of `TransferReturn` data in a couple of the more complex cases. I want to do more of that in a follow-up along with the remaining few test cases. The multi-layered stuff gets messy real quick because each struct serializes the next layer in its `return_data` or `recipient_data`. I want to spend some time coming up with a nicer way to handle that stuff but I figure it's a low priority for the moment.